File: python/src/Drawer.py
from time import sleep

'''
Drawer include:
- interface to control the drawer (opened/closed)
- interface to control the squeezer (opened/closed)
'''
class Drawer:
    
    servoInterface = None
    
    squeezerOpenedPosition = 90
    squeezerClosedPosition = 180
    
    '''
    args: {
        'squeezerServoSlot': Int
        'servoInteface': Adafruit_PCA9685.PCA9685 instance
        'arduinoInterface': Arduino instance
    }
    '''
    def __init__(self, **args):
        self.squeezerServoSlot = args['squeezerServoSlot']
        if 'servoInterface' in args:
            self.servoInterface = args['servoInterface']
        if 'arduinoInterface' in args:
            self.arduinoInterface = args['arduinoInterface']
            
    def close(self):
        self.arduinoInterface.sendCommand(
            name = 'DRAWER_GO_TO_BACK',
            expectResponse = True
        )
        
    def open(self):
        self.arduinoInterface.sendCommand(
            name = 'DRAWER_GO_TO_FRONT',
            expectResponse = True
        )
    
    def setSqueezerPosition(self, position):
        print(position)
        if self.servoInterface != None:
            self.servoInterface.set_pwm(self.squeezerServoSlot, 0, position)
        else:
            print(self.label, 'Set squeezer position to', position)
        
    def squeeze(self):
        self.setSqueezerPosition(self.squeezerOpenedPosition)
        
    def unSqueeze(self):
        self.setSqueezerPosition(self.squeezerClosedPosition)
        
    def init(self):
        self.squeeze()
        sleep(4)
        self.close()

File: node/src/Drawer.js
module.exports = class Drawer {
    constructor(params) {
        this.pwmInterface = params.pwmInterface
        this.arduinoInterface = params.arduinoInterface
        this.squeezerSlot = 7
        this.drawerOpened = false
        this.squeezerOpened = false
    }

    async goToBack() {
        if (this.squeezerOpened) {
            console.warn("> WARNING: The squeezer will be completly closed before closing the drawer")

            this.completlySqueeze()

            setTimeout(async () => {
                await this.arduinoInterface.sendCommand('DRAWER_GO_TO_BACK', [200])
            }, 700)
        } else {
            this.completlySqueeze()
            await this.arduinoInterface.sendCommand('DRAWER_GO_TO_BACK', [200])
        }
        this.drawerOpened = false
    }

    async goToFront() {
        await this.arduinoInterface.sendCommand('DRAWER_GO_TO_FRONT', [200])
        this.drawerOpened = true
    }

    toggle() {
        if (this.drawerOpened) {
            this.goToBack()
        } else {
            this.goToFront()
        }
    }
    
    closeSqueezer() {
        this.pwmInterface.setAngle(this.squeezerSlot, 175)
        this.squeezerOpened = false

        console.log('squeezer closed')
    }
    
    openSqueezer() {
        if (!this.drawerOpened) {
            console.warn("> WARNING: Can't open squeezer because the drawer is closed")
        } else {
            this.pwmInterface.setAngle(this.squeezerSlot, 90)
            this.squeezerOpened = true
            console.log('squeezer opened')
        }
    }

    completlySqueeze() {
        this.pwmInterface.setAngle(this.squeezerSlot, 180)
        console.log('COMPLETLY Squeeze')
        this.squeezerOpened = false
    }

    toggleSqueezer() {
        if (!this.squeezerOpened) {
            this.openSqueezer()
        } else {
            this.closeSqueezer()
        }
    }
}